cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java

```diff
@@ -0,0 +1,143 @@
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * perform table data migration from v2 metrics table to v3.
+ */
+public class  MetricsTableMigration {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsTableMigration.class);
+
+  private final MetricsTable v2MetricsTable;
+  private final MetricsTable v3MetricsTable;
+  private final CConfiguration cConf;
+
+  public MetricsTableMigration(MetricsTable v2MetricsTable, MetricsTable v3MetricsTable,
+                               CConfiguration cConf) {
+    this.v2MetricsTable = v2MetricsTable;
+    this.v3MetricsTable = v3MetricsTable;
+    this.cConf = cConf;
+  }
+
+  /**
+   * returns true if old metrics data is still available for transfer
+   * @return
+   */
+  public boolean isOldMetricsDataAvailable() {
+    try (Scanner scanner = v2MetricsTable.scan(null, null, null)) {
+      Row row = scanner.next();
+      return (row != null);
+    }
+  }
+
+  /**
+   * transfers data from v2 metrics table to v3 metrics table, limit on number of records is specified by
+   * maxRecordsToScan
+   * @param maxRecordsToScan - limit on number of records scanned
+   */
+  public void transferData(int maxRecordsToScan) {
+    try (Scanner scanner = v2MetricsTable.scan(null, null, null)) {
+      LOG.trace("Starting scanning for Metrics Data Migration with {} max records", maxRecordsToScan);
+      Row row;
+      int recordsScanned = 0;
+      while ((recordsScanned < maxRecordsToScan) && ((row = scanner.next()) != null)) {
+        if (recordsScanned % 10 == 0) {
+          LOG.trace("Scanned {} records in Metrics Data Migration", recordsScanned);
+        }
+
+        byte[] rowKey = row.getRow();
+        Map<byte[], byte[]> columns = row.getColumns();
+        //row-map
+        SortedMap<byte[], SortedMap<byte[], Long>> rowGauges = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+        NavigableMap<byte[], NavigableMap<byte[], Long>> rowIncrements = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+        // column-map gauges
+        List<byte[]> gaugeDeletes = new ArrayList<>();
+        // column-map increments
+        NavigableMap<byte[], Long> increments = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+        for (Map.Entry<byte[], byte[]> entry : columns.entrySet()) {
+          // column is timestamp, do a get on the new table
+          byte[] value = v3MetricsTable.get(rowKey, entry.getKey());
+          if (value == null) {
+            // we perform checkAndPut for the new value, if the value was updated by the new upcoming metrics,
+            // if it was a gauge, the processing would have deleted the entry from the old table
+            // if it was a incrment, we would perform increment in the next iteration of the migration run
+            // so its safe to skip delete on checkAndPut failure
+            if (v3MetricsTable.swap(rowKey, entry.getKey(), null, entry.getValue())) {
```

Comment:
it would be good idea to trace log if the swap fails. That would help us debug if the migration gets stuck.